src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -228,11 +230,91 @@ public ISqlConnectionFactory ConnectionFactory
+                        // Check if the token is updating already, in which case there is no need to request a new one
+                        if (this.TokenUpdateUris.ContainsKey(ownerUri)) {
+                            return false;
+                        }
+                        var requestMessage = new RefreshTokenParams
+                        {
+                            AccountId = connInfo.ConnectionDetails.GetOptionValue("azureAccount", string.Empty),
+                            TenantId = connInfo.ConnectionDetails.GetOptionValue("azureTenantId", string.Empty),
+                            Provider = "Azure",
+                            Resource = "SQL",
+                            Uri = ownerUri
+                        };
+                        if (requestMessage.TenantId == null || requestMessage.TenantId == String.Empty)
+                        {
+                            Logger.Write(TraceEventType.Error, "No tenant in connection details when refreshing token for connection {ownerUri}");
+                            return false;
+                        }
+                        if (requestMessage.AccountId == null || requestMessage.AccountId == String.Empty)
+                        {
+                            Logger.Write(TraceEventType.Error, "No accountId in connection details when refreshing token for connection {ownerUri}");
+                            return false;
+                        }
+                        try
+                        {
+                            this.TokenUpdateUris.TryAdd(ownerUri, true);
+                            await this.ServiceHost.SendEvent(RefreshTokenNotification.Type, requestMessage);
+                        }
+                        catch (Exception ex)
+                        {
+                            Logger.Write(TraceEventType.Error, "Failed to send RefreshTokenNotification " + ex.Message);
+                            return false;
+                        }
+                        return true;
+                    }
+                    else
+                    {
+                        return false;
+                    }
+                }
+            }
+            else
+            {
+                Logger.Write(TraceEventType.Error, "Failed to find connection when refreshing token");
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Requests an update of the azure auth token
+        /// </summary>
+        /// <param name="refreshToken">The token to update</param>
+        /// <returns>true upon successful update, false if it failed to find 
+        /// the connection</returns>
+        internal void UpdateAuthToken(TokenRefreshedParams tokenRefreshedParams)
+        {
+            if (!this.TryFindConnection(tokenRefreshedParams.Uri, out ConnectionInfo connection))
+            {
+                Logger.Write(TraceEventType.Error, "Failed to find connection when refreshing token");

```

Comment:
`when updating refreshed token for URI {tokenRefreshedParams.Uri}` - log messages should be unique enough to tell where exactly they came from (currently this is duplicated in a different context on line 297) and contain information to help with debugging.